common/ipmi/oem_1s_handler.c

```diff
@@ -0,0 +1,864 @@
+		return;
+	}
+
+	if (msg->data_len == 3) {
+		if ((writeLen == 0) && (readLen == 0)) {
+			ret = peci_ping(addr);
+			msg->data[0] = ret;
+			msg->data_len = 1;
+			msg->completion_code = CC_SUCCESS;
+			return;
+		} else {
+			msg->completion_code = CC_INVALID_LENGTH;
+			return;
+		}
+	} else {
+		if ((cmd == PECI_GET_DIB_CMD) || (cmd == PECI_GET_TEMP0_CMD)) {
+			readBuf = (uint8_t *)malloc(sizeof(uint8_t) * (readLen + 1));
+		} else {
+			readBuf = (uint8_t *)malloc(sizeof(uint8_t) * readLen);
+		}
+		writeBuf = (uint8_t *)malloc(sizeof(uint8_t) * writeLen);
+		if ((readBuf == NULL) || (writeBuf == NULL)) {
+			printk("PECI access util buffer alloc fail\n");
+			if (writeBuf != NULL) {
+				free(writeBuf);
+			}
+			if (readBuf != NULL) {
+				free(readBuf);
+			}
+			msg->completion_code = CC_OUT_OF_SPACE;
+			return;
+		}
+		memcpy(&writeBuf[0], &msg->data[4], writeLen);
+
+		ret = peci_write(cmd, addr, readLen, readBuf, writeLen, writeBuf);
+		if (ret) {
+			if (writeBuf != NULL) {
+				free(writeBuf);
+			}
+			if (readBuf != NULL) {
+				free(readBuf);
+			}
+			msg->completion_code = CC_UNSPECIFIED_ERROR;
+			return;
+		}
+		memcpy(&msg->data[0], &readBuf[0], readLen);
+		if ((cmd != PECI_GET_DIB_CMD) && (cmd != PECI_GET_TEMP0_CMD)) {
+			if (msg->data[0] != PECI_CC_RSP_SUCCESS) {
+				msg->data[0] = (msg->data[0] == 0xf9) ? PECI_CC_ILLEGAL_REQUEST :
+									      msg->data[0];
+				memset(&msg->data[1], 0xff, readLen - 1);
+			}
+		}
+		if (writeBuf != NULL) {
+			free(writeBuf);
+		}
+		if (readBuf != NULL) {
+			free(readBuf);
+		}
+		msg->data_len = readLen;
+		msg->completion_code = CC_SUCCESS;
+		return;
+	}
+}
+#endif
+
+#ifdef CONFIG_JTAG
+__weak void OEM_1S_SET_JTAG_TAP_STA(ipmi_msg *msg)
+{
+	if (msg->data_len != 2) {
+		msg->completion_code = CC_INVALID_LENGTH;
+		return;
+	}
+	uint8_t tapbitlen, tapdata;
+
+	tapbitlen = msg->data[0];
+	tapdata = msg->data[1];
+	jtag_set_tap(tapdata, tapbitlen);
+
+	msg->data_len = 0;
+	msg->completion_code = CC_SUCCESS;
+	return;
+}
+
+__weak void OEM_1S_JTAG_DATA_SHIFT(ipmi_msg *msg)
+{
+	uint8_t lastidx;
+	uint16_t writebitlen, readbitlen, readbyte, databyte;
+
+	writebitlen = (msg->data[1] << 8) | msg->data[0];
+	databyte = (writebitlen + 7) >> 3;
+	readbitlen = (msg->data[3 + databyte] << 8) | msg->data[2 + databyte];
+	readbyte = (readbitlen + 7) >> 3;
+	lastidx = msg->data[4 + databyte];
+
+	if (msg->data_len != (5 + databyte)) {
+		msg->completion_code = CC_INVALID_LENGTH;
+		return;
+	}
+
+	uint8_t shiftdata[databyte], receivedata[readbyte];
+	memset(shiftdata, 0, databyte);
+	memset(receivedata, 0, readbyte);
+	memcpy(shiftdata, &msg->data[2], databyte);
+	jtag_shift_data(writebitlen, shiftdata, readbitlen, receivedata, lastidx);
+
+	memcpy(&msg->data[0], &receivedata[0], readbyte);
+	msg->data_len = readbyte;
+	msg->completion_code = CC_SUCCESS;
+	return;
+}
+
+#ifdef ENABLE_ASD
+__weak void OEM_1S_ASD_INIT(ipmi_msg *msg)
+{
+	if (msg->data_len != 1) {
```

Comment:
msg == NULL check?

Reply:
We have fixed it and done the forced update.
Please review it. Thanks.